File: solutions/c/beer-song/1/src/beer_song.c
#include "beer_song.h"
#include <string.h>
#include <stdio.h>

void verse(char *buffer, unsigned int verse_number)
{
    if (verse_number >= 2)
    {
        sprintf(buffer, "%d bottles of beer on the wall, %d bottles of beer.\n"
                        "Take one down and pass it around, %d %s of beer on the wall.\n",
                verse_number, verse_number, verse_number - 1, (verse_number - 1 == 1 ? "bottle" : "bottles"));
    }
    else if (verse_number == 1)
    {
        sprintf(buffer, "1 bottle of beer on the wall, 1 bottle of beer.\n"
                        "Take it down and pass it around, no more bottles of beer on the wall.\n");
    }
    else
    {
        sprintf(buffer, "No more bottles of beer on the wall, no more bottles of beer.\n"
                        "Go to the store and buy some more, 99 bottles of beer on the wall.\n");
    }
}
void sing(char *buffer, unsigned int starting_verse_number,
          unsigned int ending_verse_number)
{
    strcpy(buffer, "");
    for (int i = (int)starting_verse_number; i >= (int)ending_verse_number; i--)
    {
        char verse_buffer[200] = "";
        verse((char *)&verse_buffer, i);
        strcat(buffer, (char *)&verse_buffer);
        if (i > (int)ending_verse_number)
            strcat(buffer, "\n");
    }
}

File: solutions/php/beer-song/1/BeerSong.php
<?php

declare(strict_types=1);

class BeerSong
{
    public function verse(int $number): string
    {
        if ($number > 2){
            $oneLess = $number - 1;
            return <<<"END"
            $number bottles of beer on the wall, $number bottles of beer.
            Take one down and pass it around, $oneLess bottles of beer on the wall.

            END;
        } else if ($number == 2) {
            $oneLess = $number - 1;
            return <<<"END"
            $number bottles of beer on the wall, $number bottles of beer.
            Take one down and pass it around, $oneLess bottle of beer on the wall.

            END;
        } else if ($number == 1) {
            return <<<"END"
            $number bottle of beer on the wall, $number bottle of beer.
            Take it down and pass it around, no more bottles of beer on the wall.

            END;
        } else {
            return <<<'END'
            No more bottles of beer on the wall, no more bottles of beer.
            Go to the store and buy some more, 99 bottles of beer on the wall.
            END;
        }
    }

    public function verses(int $start, int $finish): string
    {
        return implode("\n", array_map(array($this, 'verse'), range($start, $finish)));
    }

    public function lyrics(): string
    {
        return $this->verses(99, 0);
    }
}
